File: graph-dynamic/src/main/scala/com/cc/graph/gep/workflow/WorkFlow.scala
package com.cc.graph.gep.workflow

import java.util.concurrent.{ ThreadLocalRandom => TLRandom }
import scala.annotation.tailrec
import scala.collection.mutable.ListBuffer
import com.cc.graph.Conf
import com.cc.graph.algorithm.Modularity
import com.cc.graph.base.Graph
import com.cc.graph.gep.Chromosome
import com.cc.graph.gep.Gene
import com.cc.graph.gep.Population
import com.cc.graph.mo.NSGAII
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Await
import scala.concurrent.duration._
import com.cc.graph.algorithm.NMI

case class Result(bests: List[Chromosome])

class WorkFlow {

  selection: SelectionStrategy =>

  import Conf.Gep._

  /**
   * run the init population<br/>
   * the related props are from application.conf
   */
  def run(graphs: Graph*): Result = {

    val graphArray = graphs.toArray

    if (graphArray.isEmpty) {
      Result(Nil)
    } else {
      val chroms = ListBuffer[Chromosome]()

      // when timestamp is 0
      val firstGraph = graphs.head
      println("timestamp-0")
      val best = runTimestamp0(graphs.head, generationNum)
      chroms += best

      //when timestamp is larger than 0
      val tail = graphs.tail
      var lastBest = best

      for (index <- 1 until graphArray.size) {
        println("timestamp-" + index)
        val current = runTimestampN(graphArray(index), generationNum, lastBest)
        chroms += current
        lastBest = current
      }
      Result(chroms.toList)
    }
  }

  private def runTimestamp0(graph: Graph, maxGenerationNum: Int): Chromosome = {
    val initPopulation = Population.generate(graph, populationSize)
    val lastPop = innerTimestamp0(initPopulation, graph, generationNum)
    // get the best result back
    selection.choose(lastPop.chromosomes, graph, 0).best.get
  }

  @tailrec
  private def innerTimestamp0(lastPopulation: Population, graph: Graph, maxGenerationNum: Int): Population = {
    println(lastPopulation.generationNum + "/" + maxGenerationNum)
    if (lastPopulation.generationNum >= maxGenerationNum) {
      lastPopulation
    } else {

      val mutatedChromsFutures = for (chrom <- lastPopulation.chromosomes) yield {
        // use Future for concurrency
        Future(operateChromosome(chrom, graph))
      }

      val mutatedChroms = mutatedChromsFutures.map(f => Await.result(f, 10 seconds))

      // remain one positive for the best
      val choosedChroms = selection.choose(lastPopulation.chromosomes ++ mutatedChroms, graph, lastPopulation.chromosomes.size - 1)

      // choose the best one
      val theRemainedOne = choosedChroms.best.getOrElse(Chromosome.generate(graph))
      println(Modularity.compute(theRemainedOne.toCommunityStyle, graph).sum)

      val newPopulation = Population(choosedChroms.selected :+ theRemainedOne, lastPopulation.generationNum + 1)
      innerTimestamp0(newPopulation, graph, maxGenerationNum)
    }
  }

  private def runTimestampN(graph: Graph, maxGenerationNum: Int, lastTimestampCommunities: Chromosome): Chromosome = {
    val initPopulation = Population.generate(graph, populationSize)
    val lastPop = innerTimestampN(initPopulation, graph, generationNum, lastTimestampCommunities)
    val levels = NSGAII.fastNondominatedSort(lastPop.chromosomes, graph, lastTimestampCommunities)
    val moCache = scala.collection.mutable.Map[Chromosome, Double]()
    val nmiCache = scala.collection.mutable.Map[Chromosome, Double]()
    val moSum = levels(0)._2.map(e => moCache.getOrElseUpdate(e, Modularity.compute(e.toCommunityStyle, graph).sum)).sum
    val nmiSum = levels(0)._2.map(e => nmiCache.getOrElseUpdate(e, NMI(e.toCommunityStyle, lastTimestampCommunities.toCommunityStyle))).sum
    levels(0)._2.maxBy(chrom => moCache.getOrElseUpdate(chrom, Modularity.compute(chrom.toCommunityStyle, graph).sum) / moSum + nmiCache.getOrElseUpdate(chrom, NMI(chrom.toCommunityStyle, lastTimestampCommunities.toCommunityStyle)) / nmiSum)
  }

  @tailrec
  private def innerTimestampN(lastPopulation: Population, graph: Graph, maxGenerationNum: Int, lastTimestampCommunities: Chromosome): Population = {
    println(lastPopulation.generationNum + "/" + maxGenerationNum)
    if (lastPopulation.generationNum >= maxGenerationNum) {
      lastPopulation
    } else {
      val p = lastPopulation.chromosomes
      val length = p.size
      val qFutures = for (chrom <- p) yield {
        // use Future for concurrency
        Future(operateChromosome(chrom, graph))
      }
      val q = qFutures.map(f => Await.result(f, 10 seconds))
      // get the mixed chroms(twice length of the original one)
      val mixed = p ++ q
      //      println("mixed size:" + mixed.size)
      val levels = NSGAII.fastNondominatedSort(mixed, graph, lastTimestampCommunities)
      //      println("levels count:" + levels.map(_._2.size).sum)
      //      println("levels length:" + levels.length)
      //      println("length:" + length)
      var rest = length
      var i = 0
      val buffer = ListBuffer[Chromosome]()
      while (rest > 0 && rest - levels(i)._2.size > 0) {
        val in = levels(i)._2
        buffer ++= in
        i += 1
        rest = rest - in.size
      }
      if (rest > 0) {
        val in = levels(i)._2
        val sorted = NSGAII.crowdingDistanceAssignment(in, graph, lastTimestampCommunities).sortBy(_._2).map(_._1)
        buffer ++= sorted.take(rest)
      }
      innerTimestampN(Population(buffer.toVector, lastPopulation.generationNum + 1), graph, maxGenerationNum, lastTimestampCommunities)
    }
  }

  /**
   * operate chromosome(mutation)<br>
   * thread safe
   */
  private def operateChromosome(chrom: Chromosome, graph: Graph): Chromosome = {
    val random = TLRandom.current()
    val ls = ListBuffer[Gene]()
    ls ++= chrom.genes
    if (ls.size == 1) {
      doGeneSplitOff(ls)
    }
    if (random.nextDouble() < geneMove) {
      doGeneMove(ls)
    }
    if (random.nextDouble() < geneExchange) {
      doGeneExchange(ls)
    }
    if (random.nextDouble() < geneSplitoff) {
      doGeneSplitOff(ls)
    }
    if (random.nextDouble() < geneMerge) {
      doGeneMerge(ls, graph)
    }
    Chromosome(ls: _*)
  }

  private def doGeneMove(genes: ListBuffer[Gene]): Unit = {
    val random = TLRandom.current()
    if (genes.size > 1) {
      val g1 = genes.remove(random.nextInt(genes.size))
      val g2 = genes.remove(random.nextInt(genes.size))
      val moved = Gene.move(g1, g2)
      genes ++= moved
    }
  }

  private def doGeneExchange(genes: ListBuffer[Gene]): Unit = {
    val random = TLRandom.current()
    if (genes.size > 1) {
      val g1 = genes.remove(random.nextInt(genes.size))
      val g2 = genes.remove(random.nextInt(genes.size))
      genes ++= Gene.exchange(g1, g2)
    }
  }

  private def doGeneSplitOff(genes: ListBuffer[Gene]): Unit = {
    val geneOfMaxSize = genes.reduce((g1, g2) => if (g1.size > g2.size) g1 else g2)
    genes -= geneOfMaxSize ++= Gene.splitoff(geneOfMaxSize)
  }

  private def doGeneMerge(genes: ListBuffer[Gene], graph: Graph): Unit = {
    val random = TLRandom.current()
    val chrom = Chromosome(genes: _*)
    if (genes.size > 1) {
      val g1 = genes.remove(random.nextInt(genes.size))
      val g2 = genes.remove(random.nextInt(genes.size))
      genes += Gene.merge(g1, g2)
    }
  }
}

object WorkFlow extends App {

  //  showZacharyInit()
  //  runFootball1()
  //  run1(seperator = " ", file = "src/main/resources/football/edges.1.1.txt")

  //    runZachary()
  //  runDolphin
  runFootballSelf()

  //  val graph = Graph.load("src/main/resources/Zachary.txt", seperator = ",")
  //  val r = Modularity.compute(List(Set("9","10","15","16","19","21","23","24","25","26","27","28","29","30","31","32","33","34"),
  //    Set("1","2","3","4","5","6","7","8","11","12","13","14","17","18","20","22")), graph)
  //  println(r.sum)

  //    val graph = Graph.load("src/main/resources/football/edges.1.1.txt", seperator = " ")
  //    println(graph.getVertexes.toList.map(a => Integer.valueOf(a.id)).sorted)
  //    println(graph.getVertexes.size)

  def runFootballN() = {
    val files = List(
      "src/main/resources/football/edges.1.1.txt",
      "src/main/resources/football/edges.1.2.txt",
      "src/main/resources/football/edges.1.3.txt")
    runN(" ", files: _*)
  }

  def runFootballSelf() = {
    val files = List(
      "src/main/resources/football/edges.1.1.txt",
      "src/main/resources/football/edges.1.1.txt",
      "src/main/resources/football/edges.1.1.txt")
    runN(" ", files: _*)
  }

  def runFootball1() = {
    val r1 = run1(seperator = " ", file = "src/main/resources/football/edges.1.1.txt")
    val r2 = run1(seperator = " ", file = "src/main/resources/football/edges.1.2.txt")
    val r3 = run1(seperator = " ", file = "src/main/resources/football/edges.1.3.txt")
    println(NMI(r1.toCommunityStyle, r2.toCommunityStyle))
    println(NMI(r2.toCommunityStyle, r3.toCommunityStyle))
  }

  def runMoGraphN() = {
    val files = List(
      "src/main/resources/mo/real.t01.edges",
      "src/main/resources/mo/real.t02.edges",
      "src/main/resources/mo/real.t03.edges",
      "src/main/resources/mo/real.t04.edges",
      "src/main/resources/mo/real.t05.edges",
      "src/main/resources/mo/real.t06.edges",
      "src/main/resources/mo/real.t07.edges",
      "src/main/resources/mo/real.t08.edges",
      "src/main/resources/mo/real.t09.edges",
      "src/main/resources/mo/real.t010.edges")
    runN(" ", files: _*)
  }

  def runMoGraph1() = {
    run1(seperator = " ", file = "src/main/resources/mo/real.t01.edges")
  }

  def runZachary() = {
    val workFlow = new WorkFlow with ModularitySelection
    val graph = Graph.load("src/main/resources/Zachary.txt", seperator = ",")
    val result = workFlow.run(graph)
    val best = result.bests(0)
    println(best)
    println(Modularity.compute(best.toCommunityStyle, graph).sum)
    graph.displayCommunity(best.toCommunityStyle)
  }

  def runDolphin() = {
    val workFlow = new WorkFlow with ModularitySelection
    val graph = Graph.load("src/main/resources/Dolphin.txt", seperator = ",")
    val result = workFlow.run(graph)
    val best = result.bests(0)
    println(best)
    println(Modularity.compute(best.toCommunityStyle, graph).sum)
    graph.displayCommunity(best.toCommunityStyle)
  }

  def run1(seperator: String, file: String) = {
    val workFlow = new WorkFlow with ModularitySelection
    val graph = Graph.load(file, seperator = seperator)
    val result = workFlow.run(graph)
    val r = result.bests(0)
    println(r)
    println("node size:" + graph.getVertexes.size)
    println(Modularity.compute(r.toCommunityStyle, graph).sum)
    println(r.genes.size)
    r
  }

  def runN(seperator: String, files: String*) = {
    val workFlow = new WorkFlow with ModularitySelection
    val graphs = files.map(file => Graph.load(file, seperator = seperator))
    val result = workFlow.run(graphs: _*)
    val bests = result.bests zip graphs

    println("edge size:")
    graphs.foreach(g => println(g.getEdges.size))

    println("node size:")
    // print node size
    graphs.foreach(g => println(g.getVertexes.size))

    println("comm size:")
    // print comm size
    result.bests.foreach(e => println(e.genes.size))

    println("modurity:")
    // compute each modularity
    bests.foreach(e => println(Modularity.compute(e._1.toCommunityStyle, e._2).sum))

    println("NMI")
    // compute NMI(sliding is used to group neighbor result: List(1,2,3) => List(List(1,2),List(2,3))
    bests.map(_._1).sliding(2).foreach(es => println(NMI(es(0).toCommunityStyle, es(1).toCommunityStyle)))
  }
}

File: graph-dynamic-java/src/main/java/com/cc/graph/gep/WorkFlow.java
package com.cc.graph.gep;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;

import com.cc.graph.Conf;
import com.cc.graph.algorithm.Algorithm;
import com.cc.graph.algorithm.Modularity;
import com.cc.graph.algorithm.NMI;
import com.cc.graph.base.GraphUtil;
import com.cc.graph.base.ImmutableGraph;
import com.cc.graph.gep.selection.ModularitySelection;
import com.cc.graph.mo.NSGAII;

public class WorkFlow {

    public static final class Result {
        private final List<Population> history = new LinkedList<>();
        private Population lastPopulation;

        public List<Population> getHistory() {
            return Collections.unmodifiableList(this.history);
        }

        public Population getLastPopulation() {
            return this.lastPopulation;
        }

        public Result pushToHistory(final Population pop) {
            this.history.add(0, pop);
            return this;
        }

        public Result updateTheLast(final Population pop) {
            this.lastPopulation = pop;
            return this;
        }

        @Override
        public int hashCode() {
            final int prime = 31;
            int result = 1;
            result = prime * result + ((this.history == null) ? 0 : this.history.hashCode());
            result = prime * result
                    + ((this.lastPopulation == null) ? 0 : this.lastPopulation.hashCode());
            return result;
        }

        @Override
        public boolean equals(final Object obj) {
            if (this == obj)
                return true;
            if (obj == null)
                return false;
            if (this.getClass() != obj.getClass())
                return false;
            final Result other = (Result) obj;
            if (this.history == null) {
                if (other.history != null)
                    return false;
            } else if (!this.history.equals(other.history))
                return false;
            if (this.lastPopulation == null) {
                if (other.lastPopulation != null)
                    return false;
            } else if (!this.lastPopulation.equals(other.lastPopulation))
                return false;
            return true;
        }

    }

    /**
     * 这个方法用于扩展算法 可以在这个方法的实现中增加所需的算法
     *
     * @param graphs
     * @return
     */
    public List<Result> run(final ImmutableGraph... graphs) {
        if (graphs.length == 0) {
            return Collections.emptyList();
        }
        final List<Result> results = new ArrayList<>(graphs.length);

        Chromosome lastSolution = null;

        for (int i = 0; i < graphs.length; i++) {
            Result thisResult;
            final ImmutableGraph graph = graphs[i];
            // timestamp = 0
            if (i == 0) {
                thisResult = this.runTimestamp(graph, new Modularity(graph));
            } else {
                // timestamp > 0
                final Algorithm a1 = new Modularity(graph);
                final Algorithm a2 = new NMI(lastSolution);
                thisResult = this.runTimestamp(graph, a1, a2);

            }
            results.add(thisResult);
            lastSolution = this.getBestChromosome(thisResult.getLastPopulation(), graph);
        }
        return results;
    }

    /**
     * 在最后一代选择最好的结果 这边以模块度为最好的结果训责算法
     *
     * @param pop
     * @param graph
     * @return
     */
    private Chromosome getBestChromosome(final Population pop, final ImmutableGraph graph) {
        return ModularitySelection.instance.choose(pop.getChromosomes(), graph).best.get();
    }

    /**
     * 此方法用于初始化网络对应的种群 并传递给进行元算的innerRun方法
     * @param graph
     * @param algorithms
     * @return
     */
    private Result runTimestamp(final ImmutableGraph graph, final Algorithm... algorithms) {
        final Population initPopulation = Population.generate(graph, Conf.Gep.populationSize);
        return this.innerRun(
                new Result().updateTheLast(initPopulation).pushToHistory(initPopulation),
                Conf.Gep.generationNum, algorithms);
    }

    /**
     * 世代迭代
     * @param lastResult
     * @param maxGenerationNum
     * @param algorithms
     * @return
     */
    private Result innerRun(final Result lastResult, final int maxGenerationNum,
            final Algorithm... algorithms) {
        final Population lastPop = lastResult.lastPopulation;
        System.out.println(lastPop.getGenerationNum() + "/" + maxGenerationNum);
        if (lastPop.getGenerationNum() >= maxGenerationNum) {
            return lastResult;
        } else {
            final List<Chromosome> lastChroms = lastPop.getChromosomes();
            final List<Chromosome> mixedChroms = lastChroms.stream()
                    .map(c -> this.operateChromosome(c)).collect(Collectors.toList());
            mixedChroms.addAll(lastChroms);

            final List<List<Chromosome>> levels = NSGAII.fastNondominatedSort(mixedChroms,
                    algorithms);
            System.out.println(levels.size());
            final List<Chromosome> newChromosomes = new ArrayList<>();
            int index = 0;
            while ((lastChroms.size() - newChromosomes.size()) >= levels.get(index).size()) {
                newChromosomes.addAll(levels.get(index));
                index += 1;
            }
            final int remaining = lastChroms.size() - newChromosomes.size();
            if (remaining != 0) {
                final List<Chromosome> sorted = NSGAII.crowdingDistanceSort(levels.get(index),
                        algorithms);
                newChromosomes.addAll(sorted.subList(0, remaining));
            }
            final Population population = new Population(newChromosomes,
                    lastPop.getGenerationNum() + 1);
            return this.innerRun(lastResult.pushToHistory(population).updateTheLast(population),
                    maxGenerationNum, algorithms);
        }
    }

    /**
     * 染色体操作
     * @param chrom
     * @return
     */
    private Chromosome operateChromosome(final Chromosome chrom) {
        final ThreadLocalRandom random = ThreadLocalRandom.current();
        final List<Gene> ls = new ArrayList<Gene>(chrom.genes.size());
        ls.addAll(chrom.genes);
        if (random.nextDouble() <= Conf.Gep.geneMove) {
            this.doGeneMove(ls);
        }
        if (random.nextDouble() <= Conf.Gep.geneExchange) {
            this.doGeneExchange(ls);
        }
        if (random.nextDouble() <= Conf.Gep.geneMerge) {
            this.doGeneMerge(ls);
        }
        if (random.nextDouble() <= Conf.Gep.geneSplitoff) {
            this.doGeneSplitOff(ls);
        }
        return new Chromosome(ls);
    }

    private void doGeneMove(final List<Gene> genes) {
        final ThreadLocalRandom random = ThreadLocalRandom.current();
        if (genes.size() > 1) {
            final Gene g1 = genes.remove(random.nextInt(genes.size()));
            final Gene g2 = genes.remove(random.nextInt(genes.size()));
            final List<Gene> moved = GeneUtil.move(g1, g2);
            genes.addAll(moved);
        }
    }

    private void doGeneExchange(final List<Gene> genes) {
        final ThreadLocalRandom random = ThreadLocalRandom.current();
        if (genes.size() > 1) {
            final Gene g1 = genes.remove(random.nextInt(genes.size()));
            final Gene g2 = genes.remove(random.nextInt(genes.size()));
            genes.addAll(GeneUtil.exchange(g1, g2));
        }
    }

    private void doGeneSplitOff(final List<Gene> genes) {
        int maxPosition = 0;
        for (int i = 0; i < genes.size(); i++) {
            if (genes.get(i).size() > genes.get(maxPosition).size()) {
                maxPosition = i;
            }
        }
        final Gene g = genes.remove(maxPosition);
        genes.addAll(GeneUtil.splitoff(g));
    }

    private void doGeneMerge(final List<Gene> genes) {
        final ThreadLocalRandom random = ThreadLocalRandom.current();
        if (genes.size() > 1) {
            final Gene g1 = genes.remove(random.nextInt(genes.size()));
            final Gene g2 = genes.remove(random.nextInt(genes.size()));
            genes.add(GeneUtil.merge(g1, g2));
        }
    }

    public static void main(final String[] args) throws IOException {
        final WorkFlow workFlow = new WorkFlow();
        final ImmutableGraph graph = GraphUtil.load("src/main/resources/Zachary.txt");
        final WorkFlow.Result r = workFlow.run(graph).get(0);
        final Population pop = r.getLastPopulation();
        final List<Chromosome> cs = pop.getChromosomes();
        final List<Chromosome> ccs = new ArrayList<>(cs);
        final Modularity q = new Modularity(graph);
        final List<Double> modularities = ccs.stream().map(c -> q.compute(c))
                .collect(Collectors.toList());
        double bestModularity = -1;
        int bestIndex = 0;
        for (int i = 0; i < ccs.size(); i++) {
            final double current = modularities.get(i);
            if (current > bestModularity) {
                bestIndex = i;
                bestModularity = current;
            }
        }
        final Chromosome bestChromo = ccs.get(bestIndex);
        System.out.println("best:");
        System.out.println(bestChromo);
        System.out.println(modularities.get(bestIndex));
        graph.displayCommunity(bestChromo.toCommunityStyle());
    }
}
